presto-orc/src/main/java/com/facebook/presto/orc/DictionaryCompressionOptimizer.java

```diff
@@ -143,77 +197,107 @@ public void optimize(int bufferedBytes, int stripeRowCount)
+        removeDirectConversionCandidate(index);
+        if (directBytes.isPresent()) {
+            bufferedBytesCounter.incrementBufferedBytes(directBytes.getAsInt() - dictionaryBytes);
+            bufferedBytesCounter.incrementNonDictionaryBufferedBytes(directBytes.getAsInt());
+        }
+        return true;
+    }
+
     @VisibleForTesting
-    int convertLowCompressionStreams(int bufferedBytes)
+    int convertLowCompressionStreams(boolean tryAllStreams, int bufferedBytes)
     {
         // convert all low compression column to direct
-        for (DictionaryColumnManager dictionaryWriter : ImmutableList.copyOf(directConversionCandidates)) {
-            if (dictionaryWriter.getCompressionRatio() < DICTIONARY_MIN_COMPRESSION_RATIO) {
-                int columnBufferedBytes = toIntExact(dictionaryWriter.getBufferedBytes());
-                OptionalInt directBytes = tryConvertToDirect(dictionaryWriter, getMaxDirectBytes(bufferedBytes));
-                if (directBytes.isPresent()) {
-                    bufferedBytes = bufferedBytes + directBytes.getAsInt() - columnBufferedBytes;
-                    if (bufferedBytes >= stripeMaxBytes) {
-                        return bufferedBytes;
+        Iterator<DictionaryColumnManager> iterator = directConversionCandidates.iterator();
+        while (iterator.hasNext()) {
+            DictionaryColumnManager dictionaryWriter = iterator.next();
+            if (tryAllStreams || dictionaryWriter.getDictionaryBytes() >= dictionaryUsefulCheckColumnSizeBytes) {
+                if (dictionaryWriter.getCompressionRatio() < DICTIONARY_MIN_COMPRESSION_RATIO) {
+                    int columnBufferedBytes = toIntExact(dictionaryWriter.getBufferedBytes());
+                    OptionalInt directBytes = tryConvertToDirect(dictionaryWriter, getMaxDirectBytes(bufferedBytes));
+                    iterator.remove();
+                    if (directBytes.isPresent()) {
+                        bufferedBytes = bufferedBytes + directBytes.getAsInt() - columnBufferedBytes;
+                        if (bufferedBytes >= stripeMaxBytes) {
+                            return bufferedBytes;
+                        }
                     }
                 }
             }
         }
         return bufferedBytes;
     }
 
+    @VisibleForTesting
+    List<DictionaryColumnManager> getDirectConversionCandidates()
+    {
+        return directConversionCandidates;
+    }
+
     private void updateDirectConversionCandidates()
     {
         // Writers can switch to Direct encoding internally. Remove them from direct conversion candidates.
         directConversionCandidates.removeIf(DictionaryColumnManager::isDirectEncoded);
     }
 
+    private void removeDirectConversionCandidate(int index)
+    {
+        DictionaryColumnManager last = directConversionCandidates.get(directConversionCandidates.size() - 1);
```

Comment:
For consideration:
List.remove(idx) returns the element at this position, you can try to use one remove/set instead of get/set/remove. Of course it will fail if directConversionCandidates only have a single element.

Reply:
I started with that, but doing a check for size == 1 and element == 0. felt like lot of code. so went with this approach.